### PR TITLE
Spring.Echo: Support echoing tables

### PR DESCRIPTION
**Pull request by WatchTheFort**
### Work done

- Override `Spring.Echo` to support echoing tables
- Remove `TableEcho` function, superseded by `table.toString`

This makes debugging less frustrating, where you had to use two functions depending on the data type of the variable you wanted to echo. Tables are first printed as _\<table\>_ in the echo line, then the contents of each table are printed afterwards, in the order they were passed.

#### Test steps
- [ ] Call `Spring.Echo` and pass it a mix of table and non-table parameters. Verify the contents of the table are echoed.

**Note: no special handling for recursive tables has been added, echoing them will still crash your game.**

Note: This work will hopefully be superseded by native engine handling of printing tables, and this code will one day be removed. https://github.com/beyond-all-reason/spring/issues/1613

### CODE REVIEW

**Comment by Damgam** [comment]
image

File: common/springOverrides.lua [hunk 1]
```diff
@@ -41,3 +41,29 @@ if Spring.GetModOptions then
...
+
+		echo(unpack(args, 1, args.n))
+
+		for _, index in ipairs(tableIndexes) do
+			echo(table.toString(args[index], printOptions))
+		end
+	end
+
+	Spring.Echo = multiEcho
+end
```
Comment (salinecitrine) [inline common/springOverrides.lua:69]:
```suggestion
end

```

File: common/springOverrides.lua [hunk 2]
```diff
@@ -41,3 +41,29 @@ if Spring.GetModOptions then
...
+
+		for index = 1, args.n do
+			local value = args[index]
+
+			if type(value) == 'table' then
+				table.insert(tableIndexes, index)
+			end
+		end
+
+		echo(unpack(args, 1, args.n))
```
Comment (salinecitrine) [inline common/springOverrides.lua:61]: When I try printing out just a single table, I get output like this:

```
TABLE: 1, 2, 3
{
  1,
  2,
  3,
}
```

Apparently (I didn't know about this before) the engine [attempts to print out single tables as if they are arrays](https://github.com/beyond-all-reason/spring/blob/0787116093846d5e49f62fe3f83066e695cefcd0/rts/Lua/LuaUtils.cpp#L1253).

Given that the engine behavior is redundant and also lacking the features we already have in even the old `TableEcho`, I think we could do something like this (just some quick code to demonstrate what I'm thinking):

```suggestion
		local echoArgs = {}
		for index = 1, args.n do
			local value = args[index]

			if type(value) == 'table' then
				table.insert(tableIndexes, index)
				table.insert(echoArgs, "<table>")
			else
				table.insert(echoArgs, value)
			end
		end

		echo(unpack(echoArgs))
```

**Review by salinecitrine — COMMENTED** [review]
Aside from some weird output when passing single tables, looks good to me. Tested it in-game and didn't see anything else unexpected when using it.

> Note: no special handling for recursive tables has been added, echoing them will still crash your game.

This should work fine, since `table.toString` can handle recursive tables.

---

Also, very nice to see a "+58/−316" PR :)

**Pull request merged**